// ms-rest-js/dist/lib/util/xml.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var xml2js = tslib_1.__importStar(require("xml2js"));
function stringifyXML(obj, opts) {
    var builder = new xml2js.Builder({
        explicitArray: false,
        explicitCharkey: false,
        rootName: (opts || {}).rootName,
        renderOpts: {
            pretty: false
        }
    });
    return builder.buildObject(obj);
}
exports.stringifyXML = stringifyXML;
function parseXML(str) {
    var xmlParser = new xml2js.Parser({
        explicitArray: false,
        explicitCharkey: false,
        explicitRoot: false
    });
    return new Promise(function (resolve, reject) {
        xmlParser.parseString(str, function (err, res) {
            if (err) {
                reject(err);
            }
            else {
                resolve(res);
            }
        });
    });
}
exports.parseXML = parseXML;
